src/OSPSuite.Core/Domain/Populations/ParameterValuesCache.cs

```diff
@@ -0,0 +1,188 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using OSPSuite.Assets;
+using OSPSuite.Core.Extensions;
+using OSPSuite.Utility.Collections;
+using OSPSuite.Utility.Exceptions;
+using OSPSuite.Utility.Extensions;
+
+namespace OSPSuite.Core.Domain.Populations
+{
+   public interface IParameterValueCache
+   {
+      bool Has(string parameterPath);
+      IReadOnlyList<double> ValuesFor(string parameterPath);
+      IReadOnlyList<double> PercentilesFor(string parameterPath);
+      string[] AllParameterPaths();
+
+      /// <summary>
+      ///    Similar to <see cref="ValuesFor" /> but returns an array for R compatibility
+      /// </summary>
+      double[] GetValues(string parameterPath);
+
+      void Remove(string parameterPath);
+      void SetValues(string parameterPath, IReadOnlyList<RandomValue> newValues);
+      void SetValues(string parameterPath, IReadOnlyList<double> newValues);
+      IReadOnlyCollection<ParameterValues> AllParameterValues { get; }
+      ParameterValues ParameterValuesFor(string parameterPath);
+      ParameterValues ParameterValuesAt(int index);
+      void Add(ParameterValues parameterValues);
+      void Add(IReadOnlyCollection<ParameterValue> parameterValues);
+      void RenamePath(string oldPath, string newPath);
+      int Count { get; }
+   }
+
+   /// <summary>
+   ///    Represents a cache containing the population values. Values are stored like so:
+   ///    1 - One column per parameter path
+   ///    2 - All values for one parameter are stored in one column. That means that one individual can be read as a row
+   /// </summary>
+   public class ParameterValuesCache : IParameterValueCache
+   {
+      private readonly ICache<string, ParameterValues> _parameterValuesCache = new Cache<string, ParameterValues>(x => x.ParameterPath);
+
+      public virtual void Add(IReadOnlyCollection<ParameterValue> parameterValues) => addValues(parameterValues);
+
+      public virtual void RenamePath(string oldPath, string newPath)
+      {
+         var possibleKey = internalKeyFor(oldPath);
+         if (string.IsNullOrEmpty(possibleKey))
+            return;
+
+         var values = _parameterValuesCache[possibleKey];
+         Remove(possibleKey);
+         _parameterValuesCache.Add(newPath, values);
+      }
+
+      public virtual int Count
+      {
+         get
+         {
+            if (!_parameterValuesCache.Any())
+               return 0;
+            return _parameterValuesCache.First().Count;
+         }
+      }
+
+      private string internalKeyFor(string parameterPath)
+      {
+         if (_parameterValuesCache.Contains(parameterPath))
```

Comment:
This is the new implementation
Checking if the parameter path exists as is (always the case in PKSim)

in R, possibly the path will be used without unit even though it was loaded with unit (because we do not strop the unit)

Reply:
2 concerns regarding the `internalKEy`-function:

1. Following situation:
 - Cache contains paths `"Container|P1"` and `"Container|P [ml]"` and `AllParameterPaths()` retrieves them in this order
 - we call `internalKeyFor("Container|P")`
 - `index = Array.FindIndex(allSupportedPath, x => x.StartsWith(parameterPath))` will return the index of `"Container|P1"`
 - `string.Equals(allSupportedPath[index].StripUnit(), parameterPath))` will return false and thus no internal key will be found. But it should be `"Container|P [ml]"`

 ➡️ I think we should use e.g. `FindAll` instead of `FindIndex` and apply unit removal to all such keys.

2. I still not quite sure if I fully understand the use case and may be the example below will never occur. Following situation:
 - cache contains `"Container|P"`
 - we call `internalKeyFor("Container|P [ml]")`
 - `Array.Find(allSupportedPath, x => x.StartsWith(parameterPath))` will not find anything and thus no internal key will be found. But it should be `"Container|P"`.

 ➡️ If this use case is valid, we should do something like
```
Array.Find(allSupportedPath, x => x.StartsWith(parameterPath) || parameterPath.StartsWith(x)
...
if (... && (string.Equals(allSupportedPath[index].StripUnit(), parameterPath) || string.Equals(allSupportedPath[index], parameterPath.StripUnit()))
...
```

Reply:
>I think we should use e.g. FindAll instead of FindIndex and apply unit removal to all such keys.

Fair enough. I'll implement


>I still not quite sure if I fully understand the use case and may be the example below will never occur. Following situation:

I don't think it can occur. This whole change is about making sure we can retrieve output by path even though they were loaded with unit (from file)

in PK-Sim:  PATH WILL NEVER BE CACHED WITH UNIT. We strip the unit when we load
in R:PATH may be cached with unit. If the cache contains unit, then the user will be able to retrtrieve with unit and without unit. This makes sense because user uses quantity and should not care about the unit

HOWEVER, there is absolutely no reason for a cache that does not contains unit to be acccess with unit. I don't see a problem here and thus believe this use case in invalid

Reply:
>HOWEVER, there is absolutely no reason for a cache that does not contains unit to be acccess with unit. I don't see a problem here and thus believe this use case in invalid

OK

Reply:
Done. 

this is the code now
```
 var matchingPath = Array.FindAll(AllParameterPaths(), x => x.StartsWith(parameterPath))
            .FirstOrDefault(p => string.Equals(p.StripUnit(), parameterPath));
```

I think this cover all tests cases. I have also added more test cases to support the scenario above